09-maze_solver/progress/internal/solver/solver.go

```diff
@@ -0,0 +1,77 @@
+package solver
+
+import (
+	"fmt"
+	"image"
+	"image/gif"
+	"log/slog"
+)
+
+// Solver is capable of finding the path from the entrance to the treasure.
+// The maze has to be a RGBA image.
+type Solver struct {
+	maze   *image.RGBA
+	config config
+
+	pathsToExplore chan []image.Point
+	quit           chan struct{}
+
+	exploredPixels chan image.Point
+	animation      *gif.GIF
+
+	solution []image.Point
+}
+
+// New builds a Solver by taking the path to the PNG maze, encoded in RGBA.
+func New(imagePath string) (*Solver, error) {
+	img, err := openMaze(imagePath)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open maze image: %w", err)
+	}
+
+	return &Solver{
+		maze:           img,
+		config:         defaultColours(),
+		pathsToExplore: make(chan []image.Point),
+		quit:           make(chan struct{}),
+		exploredPixels: make(chan image.Point),
+		animation:      &gif.GIF{},
+	}, nil
+}
+
+// Solve finds the path from the entrance to the treasure.
+func (s *Solver) Solve() error {
+	entrance, err := s.findEntrance()
+	if err != nil {
+		return fmt.Errorf("unable to find entrance: %w", err)
+	}
+
+	slog.Info(fmt.Sprintf("starting at %v", entrance))
+
+	go func() {
+		// The first pixel is on the edge, the second pixel is inwards.
+		s.pathsToExplore <- []image.Point{entrance, {1, entrance.Y}}
```

Comment:
Rather than doing it this way, you can send `[]image.Point{entrance}` and then in exploring part handle the scenario where you don't know. It makes the logic work with any entrance location.

Reply:
The linked list suggestion works perfectly fine -- `&{Path: previousSteps: nil, at: entrance}`. The out-of-bounds scenario (that this was trying to prevent) doesn't need to be handled as long as the path / treasure RGBA values // palette values aren't the zero values returned by RGBAAt (or ?PaletteAt?) when the requested pixel is outside the image